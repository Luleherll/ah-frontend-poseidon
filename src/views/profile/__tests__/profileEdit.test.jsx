import { mount, shallow } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import ProfileEditPage, { ProfileEditView } from '../profileEditView';


const mockStore = configureStore([thunk]);

describe('Profile edit view', () => {
  let wrapper;
  let store;
  let instance;

  const props = {
    getPrivateDataThunk: jest.fn(),
    postDataThunk: jest.fn(() => Promise.resolve({})),
    profile: {
      profile: {
        username: 'kabanga',
        email: 'bill.twinomuhwezi@andela.com',
        bio: 'I love sports and tech',
        image: null,
        following: false,
      },
    },
  };

  const historyMock = { push: jest.fn() };

  beforeEach(() => {
    store = mockStore({});
    store.dispatch = jest.fn();
    wrapper = shallow(<ProfileEditView {...props} history={historyMock} />);
    instance = wrapper.instance();
    jest.spyOn(instance, 'handleCancelClick');
    jest.spyOn(instance, 'handleSubmit');
    jest.spyOn(instance, 'handleChange');
  });

  it('renders container with the child component successfully', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('handles click event for cancel button', () => {
    instance.handleCancelClick({
      preventDefault: () => {},
    });
    expect(instance.handleCancelClick).toBeCalled();
  });

  it('handles change event for inputs', () => {
    instance.handleChange({
      preventDefault: () => {},
      target: { name: 'username', value: 'bill' },
    });
    expect(instance.handleChange).toBeCalled();
  });

  it('handles submit event for save button', () => {
    instance.handleSubmit({
      preventDefault: () => {},
    });
    expect(instance.handleSubmit).toBeCalled();
  });

  it('renders loading alert', () => {
    const props = { getPrivateDataThunk: jest.fn() };
    wrapper = shallow(<ProfileEditView {...props} />);
    const inputFields = wrapper.find('.alert-primary');
    expect(inputFields).toHaveLength(1);
  });
});

describe('connected component', () => {
  let store;
  let wrapper;

  const props = {
    profile: {
      profile: {
        username: 'kabanga',
        email: 'bill.twinomuhwezi@andela.com',
        bio: 'I love sports and tech',
        image: null,
        following: false,
      },
    },
  };

  beforeEach(() => {
    store = mockStore({ profile: {} });
    store.dispatch = jest.fn();
    const getPrivateDataThunk = jest.fn();
    const postDataThunk = jest.fn();
    wrapper = mount(
      <Provider store={store}>
        <ProfileEditPage {...props} getPrivateDataThunk={getPrivateDataThunk} postDataThunk={postDataThunk} />
      </Provider>,
    );
  });

  it('should implement change when inputs receive data', () => {
    wrapper = mount(
      <ProfileEditView
        getPrivateDataThunk={jest.fn()}
        profile={{ profile: { username: 'Bill' } }}
      />,
    );
    const result = [{ secure_url: 'https://res.cloudinary.com/dos4j4vpc/image/upload/v1547647373/poseidon/cpih2ylhazsnidell5yl.jpg' }];
    global.cloudinary = {
      openUploadWidget: (params, cb) => {
        cb(null, result);
      },
    };
    wrapper.find('.file-upload').simulate('click');
  });
});
