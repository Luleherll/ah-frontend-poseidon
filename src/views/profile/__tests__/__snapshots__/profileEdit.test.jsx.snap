// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile edit view renders container with the child component successfully 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ProfileEditView
    getPrivateDataThunk={
      [MockFunction] {
        "calls": Array [
          Array [
            "profiles/null",
            [Function],
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    history={
      Object {
        "push": [MockFunction],
      }
    }
    postDataThunk={[MockFunction]}
    profile={
      Object {
        "profile": Object {
          "bio": "I love sports and tech",
          "email": "bill.twinomuhwezi@andela.com",
          "following": false,
          "image": null,
          "username": "kabanga",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <NavBar
          createArticle={[Function]}
          keyPress={[Function]}
          searchClick={[Function]}
          searchInput={[Function]}
        />,
        <ProfileEdit
          onChange={[Function]}
          onClick={[Function]}
          onClickChange={[Function]}
          onSubmit={[Function]}
          profile={
            Object {
              "bio": "I love sports and tech",
              "email": "bill.twinomuhwezi@andela.com",
              "following": false,
              "image": null,
              "username": "kabanga",
            }
          }
          state={null}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "createArticle": [Function],
          "keyPress": [Function],
          "searchClick": [Function],
          "searchInput": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "onChange": [Function],
          "onClick": [Function],
          "onClickChange": [Function],
          "onSubmit": [Function],
          "profile": Object {
            "bio": "I love sports and tech",
            "email": "bill.twinomuhwezi@andela.com",
            "following": false,
            "image": null,
            "username": "kabanga",
          },
          "state": null,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <NavBar
            createArticle={[Function]}
            keyPress={[Function]}
            searchClick={[Function]}
            searchInput={[Function]}
          />,
          <ProfileEdit
            onChange={[Function]}
            onClick={[Function]}
            onClickChange={[Function]}
            onSubmit={[Function]}
            profile={
              Object {
                "bio": "I love sports and tech",
                "email": "bill.twinomuhwezi@andela.com",
                "following": false,
                "image": null,
                "username": "kabanga",
              }
            }
            state={null}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "createArticle": [Function],
            "keyPress": [Function],
            "searchClick": [Function],
            "searchInput": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "onChange": [Function],
            "onClick": [Function],
            "onClickChange": [Function],
            "onSubmit": [Function],
            "profile": Object {
              "bio": "I love sports and tech",
              "email": "bill.twinomuhwezi@andela.com",
              "following": false,
              "image": null,
              "username": "kabanga",
            },
            "state": null,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
